src/util/amount.rs

```diff
@@ -0,0 +1,901 @@
+            "sat" => Ok(Denomination::Satoshi),
+            "msat" => Ok(Denomination::MilliSatoshi),
+            d => Err(ParseAmountError::UnknownDenomination(d.to_owned())),
+        }
+    }
+}
+
+/// An error during [Amount] parsing.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ParseAmountError {
+    /// Amount is negative.
+    Negative,
+    /// Amount is too big to fit in the data type.
+    TooBig,
+    /// Amount has higher precision than supported by the type.
+    TooPrecise,
+    /// Invalid number format.
+    InvalidFormat,
+    /// The denomination was unknown.
+    UnknownDenomination(String),
+}
+
+impl fmt::Display for ParseAmountError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ParseAmountError::UnknownDenomination(ref d) => {
+                write!(f, "unknown denomination: {}", d)
+            }
+            ref e => f.write_str(::std::error::Error::description(e)),
+        }
+    }
+}
+
+impl error::Error for ParseAmountError {
+    fn cause(&self) -> Option<&error::Error> {
+        None
+    }
+
+    fn description(&self) -> &'static str {
+        match *self {
+            ParseAmountError::Negative => "amount is negative",
+            ParseAmountError::TooBig => "amount is too big",
+            ParseAmountError::TooPrecise => "amount has a too high precision",
+            ParseAmountError::InvalidFormat => "invalid number format",
+            ParseAmountError::UnknownDenomination(_) => "unknown denomination",
+        }
+    }
+}
+
+// The inner type used to represent amounts.
+// This is done to more easily change the underlying type in case this is desired in the future.
+type Inner = u64;
+
+/// Type to represent Bitcoin amounts.
+#[derive(Copy, Clone, Hash)]
+pub struct Amount(Inner);
+// The Inner amount represents the number of satoshis.
+
+impl Amount {
+    /// The zero amount.
+    pub const ZERO: Amount = Amount(0);
+    /// Exactly one satoshi.
+    pub const ONE_SAT: Amount = Amount(1);
+    /// Exactly one Bitcoin.
+    pub const ONE_BTC: Amount = Amount(100_000_000);
+
+    /// Create a new Amount using [amount] as the Inner type.
+    fn from_inner(amount: Inner) -> Amount {
+        Amount(amount)
+    }
+
+    /// Create an Amount with satoshi precision and the given number of satoshis.
+    pub fn from_sat(satoshi: u64) -> Amount {
+        Amount::from_inner(satoshi)
+    }
+
+    /// Get the number of satoshis in this amount.
+    pub fn as_sat(self) -> u64 {
+        self.0
+    }
+
+    /// The maximum value of an Amount.
+    pub fn max_value() -> Amount {
+        Amount::from_inner(Inner::max_value())
+    }
+
+    // Don't use the Inner type in the methods below.
+    // Always use [Amount::from_sat] and [Amount::as_sat] instead.
+
+    /// Convert from a value expressing bitcoins to an Amount.
+    pub fn from_btc<T: IntoBtc>(btc: T) -> Result<Amount, ParseAmountError> {
+        btc.into_btc()
```

Comment:
I don't understand the name`from_btc`. Shouldn't this be called `from_nonbtc`?

Reply:
No, the `IntoBtc` trait means that you can convert the type interpreting the value as bitcoin into an Amount. This is taken from `rust-bitcoin-amount`, tbh. I could just remove it all.

So `from_btc(5.0)` makes a 5 bitcoin amount, so it's equivalent with `from_sat`.

The `IntoBtc` trait impl on `f64` or `serde_json::Number` mean "convert this number into Amount as if it was bitcoins". So `5.0.into_btc()` creates a 5 bitcoin Amount.

Reply:
Perhaps `AsBtc` was a better name, but again this is still there because I started off from the bitcoin-amount crate. I'm open to removing it, that would also remove our dependency on `serde_json`.
@apoelstra 
Let me just do that. I'll remove the trait and make this method take `f64`.

Reply:
Probably I wasn't expressing myself clearly. If `5.0.into_btc()` creates a 5 bitcoin amount that implies that the literal `5.0` is not `btc` because we first have to convert it into `btc`. How can we use a non-`btc` thing (`5.0`) as argument to `from_btc` whose name would imply that we give it a `btc` thing?

Reply:
I would like to add that `frm_btc` is totally confusing. I look at the signature, I look at the comment, and I have no idea what it is supposed to do.

Reply:
I removed the IntoBtc trait.